Elicitation and sampling fail in native mode

I have an app which uses elicitation inside a tool to request additional  data from the client. The app works as expected when running in JVM mode, but fails in native mode with the following exception: `2026-03-13 14:21:45,214 ERROR [io.quarkiverse.mcp.server.runtime.ToolMessageHandler] (executor-thread-1) Unable to call tool elicitation: io.vertx.core.json.EncodeException: Failed to encode as JSON: Class com.fasterxml.jackson.databind.ser.std.StdJdkSerializers$AtomicLongSerializer has no default (no arg) constructor (through reference chain: java.util.LinkedHashMap["id"])`

How to reproduce:
1. `git clone -b mcp-client-native-hang git@github.com:fedinskiy/reproducer.git reproducer && cd reproducer/server`
2. Build and start the server: ` mvn clean install -Pnative -Phttp && target/mcp-server-1.0.0-SNAPSHOT-runner -Dquarkus.profile=debug`
3. Initialize the server and client:
```
$ id=$(curl -i -H "Accept: application/json,text/event-stream" --data '{
  "jsonrpc" : "2.0",
  "id" : 0,
  "method" : "initialize",
  "params" : {
    "protocolVersion" : "2025-11-25",
    "clientInfo" : {
      "name" : "custom",
      "version" : "1.0"
    },
    "capabilities" : {
      "sampling" : {
        "tools" : { }
      },
      "elicitation" : {
        "form" : { },
        "url" : { }
      }
    }
  }
}
' localhost:8080/mcp | grep -i session-id --color=never | tr -d '\r' )
$ curl -H "$id" -H "Accept: application/json,text/event-stream" --data '{
  "jsonrpc" : "2.0",
  "id" : 1,
  "method" : "notifications/initialized"
}' localhost:8080/mcp
```
4. Call the tool, which uses elicitation:
```
curl -H "$id" -H "Accept: application/json,text/event-stream" --data '{
  "jsonrpc" : "2.0",
  "id" : 2,
  "method" : "tools/call",
  "params" : {
    "name" : "elicitation"
  }                    
}' localhost:8080/mcp
event: message
data: {"jsonrpc":"2.0","id":2,"error":{"code":-32603,"message":"Internal error"}}
```
5. Server logs contain the following:
```
2026-03-13 14:21:45,213 INFO  [io.quarkus.mcp.server.traffic] (vert.x-eventloop-thread-5) MCP message received [YjE2NGU3M2EtOTllOC00ZjNmLTkyZDAtOTQ4NDA2ZGU2MGZk]:

{
  "jsonrpc" : "2.0",
  "id" : 2,
  "method" : "tools/call",
  "params" : {
    "name" : "elicitation"
  }
}
2026-03-13 14:21:45,214 ERROR [io.quarkiverse.mcp.server.runtime.ToolMessageHandler] (executor-thread-1) Unable to call tool elicitation: io.vertx.core.json.EncodeException: Failed to encode as JSON: Class com.fasterxml.jackson.databind.ser.std.StdJdkSerializers$AtomicLongSerializer has no default (no arg) constructor (through reference chain: java.util.LinkedHashMap["id"])
	at io.quarkus.vertx.runtime.jackson.QuarkusJacksonJsonCodec.toString(QuarkusJacksonJsonCodec.java:167)
	at io.vertx.core.json.JsonObject.encodePrettily(JsonObject.java:1045)
	at io.quarkiverse.mcp.server.runtime.TrafficLogger.messageToString(TrafficLogger.java:21)
	at io.quarkiverse.mcp.server.runtime.TrafficLogger.messageSent(TrafficLogger.java:17)
	at io.quarkiverse.mcp.server.runtime.McpRequest.messageSent(McpRequest.java:27)
	at io.quarkiverse.mcp.server.http.runtime.StreamableHttpMcpMessageHandler$HttpMcpRequest.send(StreamableHttpMcpMessageHandler.java:572)
	at io.quarkiverse.mcp.server.runtime.ElicitationRequestImpl.lambda$send$1(ElicitationRequestImpl.java:81)
	at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:24)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
	at io.smallrye.mutiny.operators.uni.UniFailOnTimeout.subscribe(UniFailOnTimeout.java:36)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
	at io.smallrye.mutiny.operators.uni.UniOnItemConsume.subscribe(UniOnItemConsume.java:34)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:61)
	at io.smallrye.mutiny.groups.UniAwait.atMost(UniAwait.java:76)
	at io.smallrye.mutiny.groups.UniAwait.indefinitely(UniAwait.java:50)
	at io.quarkiverse.mcp.server.ElicitationRequest.sendAndAwait(ElicitationRequest.java:52)
	at io.quarkus.ts.mcp.AdvancedServer.elicitation(AdvancedServer.java:58)
	at io.quarkus.ts.mcp.AdvancedServer_ClientProxy.elicitation(Unknown Source)
	at io.quarkus.ts.mcp.AdvancedServer_elicitation_Invoker_R15SIxpVK0LZ8yNmHXHLknvjg3I.invoke(Unknown Source)
	at io.quarkus.ts.mcp.AdvancedServer_elicitation_LazyInvoker_R15SIxpVK0LZ8yNmHXHLknvjg3I.invoke(Unknown Source)
	at io.quarkiverse.mcp.server.runtime.FeatureManagerBase$FeatureMetadataInvoker.call(FeatureManagerBase.java:426)
	at io.quarkiverse.mcp.server.runtime.FeatureManagerBase$1.lambda$call$0(FeatureManagerBase.java:98)
	at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:68)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem$KnownItemSubscription.forward(UniCreateFromKnownItem.java:42)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem.subscribe(UniCreateFromKnownItem.java:23)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
	at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap.subscribe(UniOnItemOrFailureFlatMap.java:27)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
	at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:51)
	at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:110)
	at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:88)
	at io.quarkiverse.mcp.server.runtime.FeatureManagerBase$3.call(FeatureManagerBase.java:340)
	at io.quarkiverse.mcp.server.runtime.FeatureManagerBase$3.call(FeatureManagerBase.java:336)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$4(ContextImpl.java:192)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextImpl$1.execute(ContextImpl.java:221)
	at io.vertx.core.impl.WorkerTask.run(WorkerTask.java:56)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:677)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@25.0.2/java.lang.Thread.runWith(Thread.java:1487)
	at java.base@25.0.2/java.lang.Thread.run(Thread.java:1474)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:832)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:808)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Class com.fasterxml.jackson.databind.ser.std.StdJdkSerializers$AtomicLongSerializer has no default (no arg) constructor (through reference chain: java.util.LinkedHashMap["id"])
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1359)
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1506)
	at com.fasterxml.jackson.databind.SerializerProvider.findContentValueSerializer(SerializerProvider.java:824)
	at com.fasterxml.jackson.databind.ser.impl.PropertySerializerMap.findAndAddSecondarySerializer(PropertySerializerMap.java:90)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer._findAndAddDynamic(MapSerializer.java:1142)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer._findSerializer(MapSerializer.java:1271)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:805)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:763)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:719)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:34)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:503)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:342)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3437)
	at com.fasterxml.jackson.core.base.GeneratorBase.writeObject(GeneratorBase.java:422)
	at io.quarkus.vertx.runtime.jackson.JsonObjectSerializer.serialize(JsonObjectSerializer.java:27)
	at io.quarkus.vertx.runtime.jackson.JsonObjectSerializer.serialize(JsonObjectSerializer.java:24)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:503)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:342)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4938)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4146)
	at io.quarkus.vertx.runtime.jackson.QuarkusJacksonJsonCodec.toString(QuarkusJacksonJsonCodec.java:165)
	... 54 more
```

The similar error happens for samplign requests as well, to reproduce replace step 4 with the following script:
```
curl -H "$id" -H "Accept: application/json,text/event-stream" --data '{
  "jsonrpc" : "2.0",
  "id" : 2,
  "method" : "tools/call",
  "params" : {
    "name" : "sampled",
    "arguments" : {
      "question" : "What is the airspeed velocity of unladen swallow?"
    }
  }
}' localhost:8080/mcp
```

--- a/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ElicitationRequestImpl.java
+++ b/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ElicitationRequestImpl.java
@@ -78,7 +78,7 @@ public class ElicitationRequestImpl implements ElicitationRequest {
             }
             schema.put("required", required);
             JsonObject params = new JsonObject().put("message", message).put("requestedSchema", schema);
-            sender.send(Messages.newRequest(id, McpMethod.ELICITATION_CREATE.jsonRpcName(), params));
+            sender.send(Messages.newRequest(requestId, McpMethod.ELICITATION_CREATE.jsonRpcName(), params));
             return future;
         });
         if (!timeout.isNegative() && !timeout.isZero()) {

--- a/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/SamplingRequestImpl.java
+++ b/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/SamplingRequestImpl.java
@@ -125,7 +125,7 @@ public class SamplingRequestImpl implements SamplingRequest {
                 future.complete(samplingResponse);
             });
             id.set(requestId);
-            sender.send(Messages.newRequest(id, McpMethod.SAMPLING_CREATE_MESSAGE.jsonRpcName(), this));
+            sender.send(Messages.newRequest(requestId, McpMethod.SAMPLING_CREATE_MESSAGE.jsonRpcName(), this));
             return future;
         });
         if (!timeout.isNegative() && !timeout.isZero()) {

--- a/transports/http/integration-tests/src/main/java/io/quarkiverse/mcp/server/sse/it/ServerFeatures.java
+++ b/transports/http/integration-tests/src/main/java/io/quarkiverse/mcp/server/sse/it/ServerFeatures.java
@@ -3,13 +3,20 @@ package io.quarkiverse.mcp.server.sse.it;
 import jakarta.inject.Inject;
 
 import io.quarkiverse.mcp.server.BlobResourceContents;
+import io.quarkiverse.mcp.server.Elicitation;
+import io.quarkiverse.mcp.server.ElicitationRequest;
+import io.quarkiverse.mcp.server.ElicitationRequest.StringSchema;
 import io.quarkiverse.mcp.server.Prompt;
 import io.quarkiverse.mcp.server.PromptArg;
 import io.quarkiverse.mcp.server.PromptMessage;
 import io.quarkiverse.mcp.server.RequestUri;
 import io.quarkiverse.mcp.server.Resource;
+import io.quarkiverse.mcp.server.Sampling;
+import io.quarkiverse.mcp.server.SamplingMessage;
+import io.quarkiverse.mcp.server.SamplingRequest;
 import io.quarkiverse.mcp.server.TextContent;
 import io.quarkiverse.mcp.server.Tool;
+import io.smallrye.mutiny.Uni;
 
 public class ServerFeatures {
 
@@ -37,6 +44,38 @@ public class ServerFeatures {
     }
 
     public record Answer(String value) {
+    }
+
+    @Tool
+    Uni<String> samplingTool(Sampling sampling) {
+        if (sampling.isSupported()) {
+            SamplingRequest samplingRequest = sampling.requestBuilder()
+                    .setMaxTokens(100)
+                    .addMessage(SamplingMessage.withUserRole("What's happening?"))
+                    .build();
+            return samplingRequest.send().map(sr -> sr.content().asText().text());
+        } else {
+            return Uni.createFrom().item("Sampling not supported");
+        }
+    }
+
+    @Tool
+    Uni<String> elicitationTool(Elicitation elicitation) {
+        if (elicitation.isSupported()) {
+            ElicitationRequest request = elicitation.requestBuilder()
+                    .setMessage("What's your name?")
+                    .addSchemaProperty("name", new StringSchema(true))
+                    .build();
+            return request.send().map(response -> {
+                if (response.actionAccepted()) {
+                    return "Hello " + response.content().getString("name") + "!";
+                } else {
+                    return "Not accepted";
+                }
+            });
+        } else {
+            return Uni.createFrom().item("Elicitation not supported");
+        }
     }
 
 }

--- a/transports/http/integration-tests/src/test/java/io/quarkiverse/mcp/server/sse/it/ServerFeaturesTest.java
+++ b/transports/http/integration-tests/src/test/java/io/quarkiverse/mcp/server/sse/it/ServerFeaturesTest.java
@@ -1,6 +1,7 @@
 package io.quarkiverse.mcp.server.sse.it;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
@@ -9,10 +10,13 @@ import java.util.Map;
 
 import org.junit.jupiter.api.Test;
 
+import io.quarkiverse.mcp.server.ClientCapability;
+import io.quarkiverse.mcp.server.ElicitationResponse;
 import io.quarkiverse.mcp.server.test.McpAssured;
 import io.quarkiverse.mcp.server.test.McpAssured.McpStreamableTestClient;
 import io.quarkiverse.mcp.server.test.McpAssured.PromptInfo;
 import io.quarkus.test.junit.QuarkusTest;
+import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 
 @QuarkusTest
@@ -41,7 +45,7 @@ class ServerFeaturesTest {
         McpStreamableTestClient client = McpAssured.newConnectedStreamableClient();
         client.when()
                 .toolsList(p -> {
-                    assertEquals(2, p.size());
+                    assertEquals(4, p.size());
                     JsonObject schema = p.findByName("toLowerCase").inputSchema();
                     JsonObject properties = schema.getJsonObject("properties");
                     assertEquals(1, properties.size());
@@ -73,6 +77,81 @@ class ServerFeaturesTest {
                             r.contents().get(0).asBlob().blob());
                 })
                 .thenAssertResults();
+    }
+
+    @Test
+    void testSampling() throws InterruptedException {
+        McpStreamableTestClient client = McpAssured.newStreamableClient()
+                .setClientCapabilities(new ClientCapability(ClientCapability.SAMPLING, Map.of()))
+                .build()
+                .connect();
+
+        JsonObject request = client.newRequest("tools/call")
+                .put("params", new JsonObject()
+                        .put("name", "samplingTool"));
+        client.sendAndForget(request);
+
+        // The server should send a sampling request
+        JsonObject sr = client.waitForRequests(1).requests().get(0);
+        assertEquals("sampling/createMessage", sr.getString("method"));
+        Long id = sr.getLong("id");
+        JsonObject response = new JsonObject()
+                .put("jsonrpc", "2.0")
+                .put("result", new JsonObject()
+                        .put("role", "assistant")
+                        .put("model", "test-model")
+                        .put("content", new JsonObject()
+                                .put("type", "text")
+                                .put("text", "Hello from sampling.")))
+                .put("id", id);
+        client.sendAndForget(response);
+
+        JsonObject toolCallResponse = client.waitForResponse(request);
+        JsonObject toolCallResult = toolCallResponse.getJsonObject("result");
+        assertNotNull(toolCallResult);
+        assertFalse(toolCallResult.getBoolean("isError"));
+        JsonArray content = toolCallResult.getJsonArray("content");
+        assertEquals(1, content.size());
+        assertEquals("Hello from sampling.", content.getJsonObject(0).getString("text"));
+    }
+
+    @Test
+    void testElicitation() throws InterruptedException {
+        McpStreamableTestClient client = McpAssured.newStreamableClient()
+                .setClientCapabilities(new ClientCapability(ClientCapability.ELICITATION, Map.of()))
+                .build()
+                .connect();
+
+        JsonObject request = client.newRequest("tools/call")
+                .put("params", new JsonObject()
+                        .put("name", "elicitationTool"));
+        client.sendAndForget(request);
+
+        // The server should send an elicitation request
+        JsonObject er = client.waitForRequests(1).requests().get(0);
+        assertEquals("elicitation/create", er.getString("method"));
+        JsonObject params = er.getJsonObject("params");
+        assertEquals("What's your name?", params.getString("message"));
+        JsonObject schemaProperties = params.getJsonObject("requestedSchema").getJsonObject("properties");
+        assertEquals("string", schemaProperties.getJsonObject("name").getString("type"));
+
+        Long id = er.getLong("id");
+        JsonObject response = new JsonObject()
+                .put("jsonrpc", "2.0")
+                .put("result", new JsonObject()
+                        .put("action", ElicitationResponse.Action.ACCEPT.toString().toLowerCase())
+                        .put("content", new JsonObject()
+                                .put("name", "Quarkus")))
+                .put("id", id);
+        client.sendAndForget(response);
+
+        JsonObject toolCallResponse = client.waitForResponse(request);
+        JsonObject toolCallResult = toolCallResponse.getJsonObject("result");
+        assertNotNull(toolCallResult);
+        assertFalse(toolCallResult.getBoolean("isError"));
+        JsonArray content = toolCallResult.getJsonArray("content");
+        assertEquals(1, content.size());
+        assertEquals("Hello Quarkus!", content.getJsonObject(0).getString("text"));
     }
 
 }

